compiler/rustc_mir_transform/src/deduce_param_attrs.rs

```diff
@@ -11,98 +11,106 @@
-            // disable any sketchy `readonly` optimizations for now.
             | PlaceContext::NonMutatingUse(NonMutatingUseContext::RawBorrow) => {
-                self.read_only[param_index] |= DeducedReadOnlyParam::MUTATED;
+                // Whether mutating though a `&raw const` is allowed is still undecided, so we
+                // disable any sketchy `readonly` optimizations for now.
+                self.usage[i] |= UsageSummary::MUTATE;
+                self.usage[i] |= UsageSummary::CAPTURE;
             }
-            // Not mutating if the parameter is `Freeze`.
             PlaceContext::NonMutatingUse(NonMutatingUseContext::SharedBorrow) => {
-                self.read_only[param_index] |= DeducedReadOnlyParam::IF_FREEZE;
+                // Not mutating if the parameter is `Freeze`.
+                self.usage[i] |= UsageSummary::SHARED_BORROW;
+                self.usage[i] |= UsageSummary::CAPTURE;
             }
             // Not mutating, so it's fine.
-            PlaceContext::NonMutatingUse(..) => {}
+            PlaceContext::NonMutatingUse(
+                  NonMutatingUseContext::Inspect
+                | NonMutatingUseContext::Copy
+                | NonMutatingUseContext::Move
+                | NonMutatingUseContext::FakeBorrow
+                | NonMutatingUseContext::PlaceMention
+                | NonMutatingUseContext::Projection) => {}
         }
     }
 
     fn visit_terminator(&mut self, terminator: &Terminator<'tcx>, location: Location) {
-        // OK, this is subtle. Suppose that we're trying to deduce whether `x` in `f` is read-only
-        // and we have the following:
-        //
-        //     fn f(x: BigStruct) { g(x) }
-        //     fn g(mut y: BigStruct) { y.foo = 1 }
-        //
-        // If, at the generated MIR level, `f` turned into something like:
-        //
-        //      fn f(_1: BigStruct) -> () {
-        //          let mut _0: ();
-        //          bb0: {
-        //              _0 = g(move _1) -> bb1;
-        //          }
-        //          ...
-        //      }
-        //
-        // then it would be incorrect to mark `x` (i.e. `_1`) as `readonly`, because `g`'s write to
-        // its copy of the indirect parameter would actually be a write directly to the pointer that
-        // `f` passes. Note that function arguments are the only situation in which this problem can
-        // arise: every other use of `move` in MIR doesn't actually write to the value it moves
-        // from.
         if let TerminatorKind::Call { ref args, .. } = terminator.kind {
             for arg in args {
                 if let Operand::Move(place) = arg.node
-                    // We're only interested in arguments.
-                    && let Some(param_index) = self.as_param(place.local)
                     && !place.is_indirect_first_projection()
+                    && let Some(i) = self.as_param(place.local)
```

Comment:
Should we also check whether that argument's type may be passed indirectly? In case of projections in particular.